### PR TITLE
Exporters can now expose metrics through Micrometer face.

### LINKED ISSUE #18467: Provide observation context to exporters
**Is your feature request related to a problem? Please describe.**

I would like to add metrics from my own custom exporter and have them scraped along with the rest of the Zeebe metrics. However, using the default `CollectorRegistry.defaultRegistry` (or the Micrometer equivalent one) fails, as the exporter is loaded in a different class loader. This also implies that Zeebe will always rely on the default registry, which may not be true forever.

**Describe the solution you'd like**

Much like the `Context` provided to the exporter on configuration has a `Logger` member, I would like the same for metrics. I think using `Micrometer` (which is to metrics what SLF4J is to logging) would make more sense, but I'm OK with Prometheus :)

**Describe alternatives you've considered**

You can work around this by loading the `CollectorRegistry` class directly from the system class loader, e.g.:

```java
final Class<?> registryClass =
    ClassLoader.getSystemClassLoader().loadClass(CollectorRegistry.class.getName());
final Field registryField = registryClass.getDeclaredField("defaultRegistry");
final CollectorRegistry registry = (CollectorRegistry) registryField.get(null);
Counter.build().name("test").namespace("test").register(registry);
```

As you can see, this is cumbersome, and ignores all the try-catch and possible errors. It also relies on the `defaultRegistry` global, which makes it harder if we ever more to Micrometer across the board.

### PR DESCRIPTION
**Pull request by EuroLew**
## Description

In the `Context` provided to exporters a `getMeterRegistry` call is now exposed. All meters registered to this meter registry will be exposed on the /actuator/prometheus endpoint.

## Related issues

closes #18467

### CODE REVIEW

**Comment by npepinpe** [comment]
Couple things to watch out for:

- Exporters can be instantiated multiple times, but likely they will use the same underlying Prometheus collector. That means that the metrics may end up being registered multiple times by the user - is this achievable? If not, what could be the UX here to make it simple for users? 
  - Here I'm ignoring having different people use the same metric name for different things - that's possible, but not something we can do much about. At best, we should ensure that when this happens, users know what to do about it (whether through proper logging or documentation).
- Any reason to create a new meter registry per exporter context? We lose the caching of the metrics every time the exporter is re-instantiated. Could we not simply use a global meter registry? I would propose having the dist module inject it into the broker, so we can just use the same registry used by the application in general.
  - We would need to document for exporter authors that they should name space their metrics to avoid collision, or at least how to detect such things.

File: zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/repo/ExporterRepository.java
Comment (npepinpe) [inline zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/repo/ExporterRepository.java:95]: :thought_balloon: Remember to update the documentation [here](https://docs.camunda.io/docs/self-managed/concepts/exporters/#loading) after to add a note that the meter registry you get during the loading/validation phase is just an in-memory one, and will be discarded after.

File: zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExternalExporterContainerTest.java
Comment (npepinpe) [inline zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExternalExporterContainerTest.java:1]: :question: Should we pair on this? It would be good to see that it works with an externally loaded exporter to ensure there's no problems with the class path or anything like that.

File: zeebe/exporter-test/src/main/java/io/camunda/zeebe/exporter/test/ExporterTestContext.java
Comment (npepinpe) [inline zeebe/exporter-test/src/main/java/io/camunda/zeebe/exporter/test/ExporterTestContext.java:32]: :x: Please return a `SimpleMeterRegistry`. It should be a field in the context so it's always the same returned for this particular context, not one we create on the fly here.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java:1]: :x: I would only include the dependency to the Prometheus parser if we do anything more with the metric, like:

- Increment the counter and verify its value
- Check the metric type is the correct one
- Check its properties are well set (e.g. help, type, name, namespace, etc.)

If we just check for existence, then I don't think we need the dependency, and we can just look for the text in the dump and use a simple regular expression to check for it, since the format is basically:

```
metric_name [
  "{" label_name "=" `"` label_value `"` { "," label_name "=" `"` label_value `"` } [ "," ] "}"
] value [ timestamp ]
```

Meaning, for example:

```
http_requests_total{method="post",code="200"} 1027 1395066363000
```

Which as a regex would be:

```
^([a-zA-Z_:][a-zA-Z0-9_:]*)(\{.+?\})?\s+([a-zA-Z-+\.\d]+)(\s+([-\d]+))?$
```

Where match group:
1. Metric name
2. Labels (unparsed, would make it more complex to parse them)
3. Value
4. Timestamp + separator
5. Timestamp/Epoch

> [!Note]
> This is ignoring the comments with the HELP and TYPE of the metric encoded.

:question: Is there any value to asserting more properties than just the existence of the metric? If so, keep the dependency and add the assertions. If not, we drop it and just do simple text parsing to check for existence. Third option is you're not sure, we should discuss this :slightly_smiling_face:

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsTestExporter.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsTestExporter.java:16]: :question: I don't understand the comment. Is this a left over from development?

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/PrometheusActuator.java
Comment (npepinpe) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/PrometheusActuator.java:27]: :x: Not sure why this is a fully qualified name. Can't we shorten it to just `PrometheusActuator`?

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/PrometheusActuator.java
Comment (npepinpe) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/PrometheusActuator.java:29]: :question: Same as above - is this some IntelliJ thing? :smile:

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestStandaloneBroker.java
Comment (npepinpe) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestStandaloneBroker.java:74]:
```suggestion
    withProperty("zeebe.broker.gateway.enable", config.getGateway().isEnable()).
      withProperty("management.endpoint.prometheus.enabled", true).
      withProperty("management.prometheus.metrics.export.enabled", true);
```

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestStandaloneBroker.java
Comment (npepinpe) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestStandaloneBroker.java:73]: :question: Is this not enabled by default in our `application.properties` file? Do you know why it was necessary to add them here?

**Review by npepinpe — CHANGES_REQUESTED** [review]

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsTestExporter.java
Comment (EuroLew) [reply zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsTestExporter.java:16]: oops

**Comment by deepthidevaki** [comment]
Since Nicolas has already reviewed this, I will remove my assignment. Please rerequest if you need my reviews after making the changes.

File: zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/repo/ExporterRepository.java
Comment (EuroLew) [reply zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/repo/ExporterRepository.java:95]: Does the user need to know this? My understanding is if they add code in the `configure` block registering meters it will add it to the simple register, discard it, and then the second time round register it with the actual `PrometheusMeterRegistry` as intended.

File: zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExternalExporterContainerTest.java
Comment (EuroLew) [reply zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExternalExporterContainerTest.java:1]: Help with testing externally loaded exporters would be great, I'll set up a meeting

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestStandaloneBroker.java
Comment (EuroLew) [reply zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestStandaloneBroker.java:73]: it is I remember it not working without this but maybe it was due to a different test that used its own config and I assumed that was the base test config

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java
Comment (EuroLew) [reply zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java:1]: I guess I'm just not sure how much we would gain from not using the parsing library, It would help any tests we would do in the future which would require the actual values of the meters.

File: zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/repo/ExporterRepository.java
Comment (EuroLew) [reply zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/repo/ExporterRepository.java:95]: https://github.com/camunda/camunda-docs/pull/4082 documentation changes are here

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java:30]: :x: Ah, do we need to inject the registry here because the broker is not using the registry coming from Spring? Can you add a TODO to remove this here when we do the follow up of injecting the `MeterRegistry` from Spring into the broker after?

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsTestExporter.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsTestExporter.java:36]: :wrench: Can be removed since they don't override anything.

**Review by npepinpe — APPROVED** [review]
:+1: Please address comments before merging. I think we will also want to do the follow up now and inject the meter registry from the dist module into the broker, that way we can get rid of the todo in the QA test.

:x: Please fix the Java checks (seems like there's an unused dependency declared in one of the POMs).

**Comment by EuroLew** [comment]
> 👍 Please address comments before merging. I think we will also want to do the follow up now and inject the meter registry from the dist module into the broker, that way we can get rid of the todo in the QA test.
> 
> ❌ Please fix the Java checks (seems like there's an unused dependency declared in one of the POMs).

The meter registry is now injected through the system context.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java:35]: :x: This isn't required anymore with newer Java versions. You can just do `new String(actuator.metrics().body().asInputStream().readAllBytes(), StandardCharsets.UTF_8);`. 

That said, Feign is pretty smart - did you try just returning `String` in the `PrometheusActuator` instead of `Response`? I have a feeling it might just do the whole thing for you.

I would try the following, assuming Feign can convert the body by itself to `String`:

```java
@Test
  public void shouldAddMeterToExporterMetrics() {
    // given
    final var actuator = PrometheusActuator.of(zeebe);

    // when
    final var metricsPlainText = actuator.metrics();

    // then
    Assertions.assertThat(metricsPlainText.lines())
        .filteredOn(line -> line.startsWith(ExporterMetricsTestExporter.REGISTERED_COUNTER))
        .hasSize(1);
  }
```

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java:37]: :wrench: Since Java 11, you can just do `"foo\n".lines()` to read a String line by line.

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/PrometheusActuator.java
Comment (npepinpe) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/actuator/PrometheusActuator.java:31]: :x: No need, we don't deal with JSON for encoding/decoding anything.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/exporter/ExporterMetricsIT.java:31]: :x: With JUnit 5, the visibility level of classes and methods doesn't need to be public, can just be package private. Seems silly for tests, but I think as a rule of thumb, it's good to always go with the lowest visibility - you never know what someone might do with code you wrote :smile:

File: zeebe/qa/integration-tests/pom.xml
Comment (npepinpe) [inline zeebe/qa/integration-tests/pom.xml:480]: :x: I think these might not be required in the end, if you apply suggestions/changes.

**Review by npepinpe — APPROVED** [review]
:+1: Looks good!

Pre-approving because I don't think we need another review, but don't hesitate to ask for one if you're unsure. Do address all comments before merging though :slightly_smiling_face:

**Pull request merged**
